snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,517 @@
+
+* Goal 3: Implement snapshotting interface for Amazon EBS.
+
+* Goal 4: Implement snapshotting interface for GCE PDs.
+
+### Feature Roadmap
+
+Major features, planned for the first version:
+
+* On demand snapshots
+
+    * API to create new snapshots
+
+    * API to list snapshots available to the user
+
+    * API to delete existing snapshots
+
+    * API to create a new persistent volume with persistent volume claim from a snapshot
+
+### Future Features
+
+The features that are not planned for the first version of the API bus should be considered in future versions:
+
+* Creating snapshots
+
+    * Scheduled and periodic snapshots
+
+    * Aplication initiated on-demand snapshot creation
+
+    * Support snapshot per PVC, pod or StatefulSet
+
+    * Support snapshots for non-cloud storage volumes (plugins might require actions to be triggered from the node)
+
+    * Support application-consistent snapshots (coordinate distributed snapshots across multiple volumes)
+
+    * Enable to create a pod/statefulsets with snapshots
+
+* List snapshots
+
+    * Enable to get the list of all snapshots for a specified persistent volume
+
+    * Enable to get the list of all snapshots for a pod/StatefulSet
+
+* Delete snapshots
+
+    * Enable to automatic garbage collect older snapshots when storage is limited
+
+* Quota management
+
+    * Enable to set quota for limiting how many snapshots could be taken and saved
+
+    * When quota is exceeded, delete the oldest snapshots automatically
+
+## Requirements
+
+### Performance
+
+* Time SLA from issuing a snapshot to completion:
+
+* The period we are interested is the time between the scheduled snapshot time, the time the snapshot is cut (it is safe to write to the snapshotted volume again) and the time the snapshot is finished uploading to its storage location.
+
+* This should be on the order of a few minutes.
+
+### Reliability
+
+* Data corruption
+
+    * Though it is generally recommended to stop application writes before executing the snapshot command, we will not do this automatically for several reasons:
+
+        * GCE and Amazon can create snapshots while the application is running.
+
+        * Stopping application writes cannot be done from the master and varies by application, so doing so will introduce unnecessary complexity and permission issues in the code.
+
+        * Most file systems and server applications are (and should be) able to restore inconsistent snapshots the same way as a disk that underwent an unclean shutdown.
+
+    * The data consistency would be best-effort only: e.g., call fsfreeze prior the snapshot on filesystems that support it.
+
+    * There are several proposed solutions that would enable the users to specify the action to perform prior/after the
+    snapshots: e.g. use pod annotations.
+
+* Snapshot failure
+
+    * Case: Failure during external process, such as during API call or upload
+
+        * Log error, do not attempt to retry
```

Comment:
Yes, we prefer not to retry because normally snapshot creation failed for a good reason, retry will most likely fail again. Also user might find a snapshot is finally taken long after the request is sent. 
When controller crash-loops, the system might try to take a snapshot again. But I think if we make sure the snapshot name is the same, the volume provider should return an error indicating the snapshot is already taken.